pinot-common/src/main/java/org/apache/pinot/common/metadata/segment/SegmentPartitionMetadata.java

```diff
@@ -48,6 +53,21 @@ public SegmentPartitionMetadata(
       @Nonnull @JsonProperty("columnPartitionMap") Map<String, ColumnPartitionMetadata> columnPartitionMap) {
     Preconditions.checkNotNull(columnPartitionMap);
     _columnPartitionMap = columnPartitionMap;
+    _uploadedSegmentPartitionId = -1;
+  }
+
+  /**
+   * Constructor for the class.
+   *
+   * @param columnPartitionMap Column name to ColumnPartitionMetadata map.
+   */
+  @JsonCreator
+  public SegmentPartitionMetadata(
+      @Nullable @JsonProperty("columnPartitionMap") Map<String, ColumnPartitionMetadata> columnPartitionMap,
+      @Nullable @JsonProperty(value = "uploadedSegmentPartitionId", defaultValue = "-1")
```

Comment:
@rohityadav1993 : we could just upload segments with the appropriate name format right? Can you add some reasoning in the PR Description about why that doesn't work?

Reply:
Updated the description with an Alternatives section. There are assumptions in multiple places in Pinot codebase when using LLCSegmentName which require a much more complex handling of uploaded segments from realtime segments.

Reply:
there are a few places that directly get columnPartitionMap and iterate it to get partitionId (you can find those places by searching usage of method `getColumnPartitionMap()` in this class), and this new field might fail those places.

Reply:
@klsince, I validated that `getColumnPartitionMap()` is being called from broker code, which is needed for query routing. This should still hold good if partition column is configured for the table.

Adding additional metadata field for parititionId is to make sure the uploaded segments are placed together with the stream partitions assigned to hosts by controller.

Reply:
I see one usage in `SegmentAssignmentUtils.getPartitionId()` used by `ReplicaGroupSegmentAssignmentStrategy` for assigning offline segments.

I think we can fail fast here if column partition and uploadedSegmentPartitionId both are provided as `uploadedSegmentPartitionId` is more for keeping the assignment same as stream partitions for upserts but it does not make sense for an offline table.

Open to better name suggestion for `uploadedSegmentPartitionId` in case it sounds misleading.

Reply:
I might have missed it, but how to configure `SegmentPartitionConfig` in TableConfig for tables that allow to upload segments built and partitioned externally? and segments might be partitioned on multi-columns with hash functions not supported by Pinot? For such table, will all of its segments get uploaded after being built/partitioned externally, or a mix of segments from Realtime ingestion and segments being built externally?

Another place that might be broken with this new field `uploadedSegmentPartitionId` is in MergeRollupTaskGenerator around L430.

Reply:
> I might have missed it, but how to configure SegmentPartitionConfig in TableConfig for tables that allow to upload segments built and partitioned externally?

We don't need to configure the table similar to how we don't need to do for realtime stream ingestion for upsert tables.

Providing some more context for why the change is needed
There are two scenarios where data partitioning comes to play:
1. Query routing: [[docs](https://docs.pinot.apache.org/operators/operating-pinot/tuning/routing#data-ingested-partitioned-by-some-column)] Data partitioning is not a requirement here but a good optimization.
2. Segment assignment:
    a. If the data is partitioned on a single column and with a Pinot supported algorithm, we configure the table as:
    ```
    ...
    "tableIndexConfig": {
      ...
      "segmentPartitionConfig": {
        "columnPartitionMap": {
          "memberId": {
            "functionName": "Modulo",
            "numPartitions": 3
          }
        }
      },
      ...
    },
    ``` 
    **Partitioning for upsert tables**:
    Consuming segment assignement: The stream is always externally partitioned (either on PK or other field which can still ensure the PKs are all part of the same partition) and does not need to use one of Pinot supported algorithms. `segmentPartitionConfig` need not be set for the upsert table either. Each `LLCSegmentName` contains a partitionId substring which is derived from the streams partitionId. When assigning a segment to instance, we get the partition id by parsing the LLCSegmentName in `SegmentUtils.getRealtimeSegmentPartitionId`.
    
    **Uploaded segment assignment**: Uploaded segments are not generated with LLCSegmentName convention. The only way to specify partitioning info is via `segmentPartitionConfig` via table config which is not possible if the stream is using custom partitioning.
    If one wants to backfill/uplaod segment to such custom partitioned stream, the uploaded segment must provide the partitionId so the segmentAssignement can put the segments in the same instances as the consuming segments of the same partition un pusert table.

Reply:
make sense. thanks for the explanation.

Another concern is the comparison logic (as in ConcurrentMapPartitionUpsertMetadataManager) to break tie is not be deterministic for segments not in LLC naming pattern. While processing segments not in LLC name pattern, segments added first are favored, but we can not control which segments get processed first next time when server restarts.

Reply:
I see you have proposed in description that we can use a new naming pattern for uploaded segment. Similarly, I'd propose something a bit simpler: `uploaded_{tableName}__{partitionId}__{creationTime}`, with 3 fields as separated by `__` (double underscore), in order to provide {partitionId} info for segment assignment and help break tie with {creationTime}.

The `uploaded_` prefix can be anything or none, but better to have it to lookup uploaded segments easily, e.g. MergeRollup task names the segments it generates with a prefix `merged_` when uploading them to the table.

Reply:
> While processing segments not in LLC name pattern, segments added first are favored, but we can not control which segments get processed first next time when server restarts.

We have segment status as `UPLOADED`, we can consider comparing create/upload time to do resolution.

> uploaded_{tableName}__{partitionId}__{creationTime}

Do we enforce naming conventions? I only see `LLCSegmentName`. I can add one for uploaded segment name for realtime table segment but wasn't prefering this approach if this is not something strictly enforced.

Additionally, we will have to add a `SegmentNameGenerator` implementation so that the user workload generating and uploading segments can create these segments.

Reply:
I actually dropped {sequenceId} intentionally, otherwise, the segment name becomes LLC format, and would cause the complexities you mentioned in another alternative proposal, like the checks on start/end offsets. In fact, the endOffset of latest LLC segment (the one with max seqId) is used to resume stream consumption, and we probably don't want to affect that logic with uploaded segments.

> Do we enforce naming conventions? ... ... add a SegmentNameGenerator implementation ...

I don't think so. We may need one just for uploading segments to upsert tables, due to its special requirements on partition id and to break comparison ties. And good point for the need of a new name generator.

btw, from what I learnt while reviewing this PR, there seems a design choice for RT segments that the {partitionId} encoded in the segment name is the source of truth, rather than the one kept in ZK metadata as there might no partition info at all in ZK metadata. So following on that design principle, I'd prefer to keep encoding the {partitionId} in segment name for uploaded segments as well. In this way, we can 1) avoid the cost of reading ZK metadata every time we need partitionId for the uploaded segments; 2) avoid changes on persistent segment metadata in ZK or on disk, which might make things a bit easier when to consider upgrade/downgrade.

Reply:
Summarising the chat discussion with @klsince here:
Plan to use naming convention: `uploaded_{tableName}_{partitionId}_{sequenceId}_{creationTime}`
1. The sequenceId is needed for scenarios where bulk data can be loaded per day and creation time is constant. e.g. backfilling data per day(creation time last second of the day) and each partition will have multiple segments each differentiated by sequenceId.
2. `_` delimiter is used to avoid LLCSegment like treatment to uploaded segments.
Updating the PR accordingly.